### ISSUE TITLE
Comprehensive Code Review: Architecture, Correctness & Performance

### ISSUE THREAD

**Issue by Jairaj1234-dancer** [sub-surface/hexgo#1]
# Comprehensive Code Review — HexGo

**Reviewer:** @Jairaj1234-dancer
**Scope:** Full codebase (~3,700 LoC across 13 Python files + dashboard HTML)
**Date:** 2026-03-31

---

## Executive Summary

HexGo is an AlphaZero-style self-play training system for Hexagonal Connect-6 on an infinite grid. The codebase is well-structured, has meaningful research-grounded design choices (D6 symmetry augmentation, hex-masked convolutions, ZOI pruning, TD-lambda targets), and ships a working end-to-end pipeline from self-play through training to a live dashboard. That said, there are correctness bugs, security issues, performance bottlenecks, and architectural gaps that should be addressed. Below is a detailed review organized by severity.

---

### 1. `server.py:224` — `exec()` on config file is a remote code execution vulnerability

```python
# server.py:224
ns: dict = {}
exec(CONFIG_FILE.read_text(encoding="utf-8"), ns)
```

Both `api_config()` (GET) and `api_config_write()` (POST) use `exec()` to read `config.py`. If the dashboard is ever exposed beyond localhost or if `config.py` is modified maliciously, this executes arbitrary Python. The POST endpoint at line 229 also `exec()`s before validating, meaning a race between a malicious config write and the next read could execute injected code.

**Recommendation:** Parse `config.py` with `ast.literal_eval()` or store config as JSON. At minimum, restrict `exec()` with `{"__builtins__": {}}` as the globals.

### 2. `inference.py:281-282` — CUDA Graph value extraction reads wrong indices

```python
val_idxs = [s for s, e, r, m, k in request_slices]
values  = self._graph_val[val_idxs].detach().float().cpu().numpy()
```

`val_idxs` collects `start` indices from `request_slices`, which are row offsets into the move-expanded batch — not the value-per-request indices. The value head produces one scalar per *request* (not per move), so `val_idxs` should be `range(len(request_slices))`. The same bug exists in the eager path (line 298). This means multi-move requests get the wrong value in CUDA Graph mode.

**Recommendation:** Change to `values = self._graph_val[:len(request_slices)]` (and same for the eager path).

### 3. `train.py:475` — Auxiliary losses recompute trunk unnecessarily

```python
# train.py:475
f_aux = net.trunk(boards_aux)
```

The trunk features `f` are already computed at line 403. Yet auxiliary losses re-encode `boards_aux` (which are the *same boards* just re-stacked from the same batch items) and run `net.trunk()` again. This doubles trunk compute for every batch that has aux labels.

**Recommendation:** Index into the existing `f` tensor: `f_aux = f[aux_indices]` where `aux_indices = [i for i, b in enumerate(batch) if "own_label" in b]`.

### 4. `mcts.py:78` — `_expand` assigns child `player` incorrectly for 1-2-2 rule

```python
node.children = [Node(move=m, parent=node, prior=prior, player=game.current_player)
                 for m in moves]
```

In the pure MCTS path (no net), `_expand` creates children with `player=game.current_player`. But after the parent's move is made via `game.make()`, `current_player` may or may not have flipped depending on whether it's the first or second placement in a 2-tile turn. The children's `player` field should reflect the player who *will move from that child state*, not who is about to move before expansion. This is correct for single-placement turns but can mislabel during the second placement of a 2-tile turn.

**Impact:** Affects pure-MCTS rollouts (not net-guided). The `_backprop` sign-flip relies on `node.player` being accurate.

### 5. `tune.py:33` — Hardcoded Windows Python path

```python
PYTHON = r"C:\Program Files\Python312\python.exe"
```

This will fail on any non-Windows machine (macOS, Linux, WSL). Should use `sys.executable` like `server.py` does.

---

### 6. `train.py:410-442` — Policy loss computed sample-by-sample in a Python loop

The policy loss iterates over each batch item individually (`for i, item in enumerate(batch)`), calling `net.policy_logit()` per item with variable-length move sets. This negates GPU parallelism — the policy head runs B separate forward passes instead of one batched call. On a 64-item batch with ~20 moves each, this is ~1,280 sequential small GPU calls.

**Recommendation:** Pad move planes to a fixed max count per batch, run one batched forward pass, then mask invalid entries before computing the loss. This alone could provide a 5-10x training speedup.

### 7. `game.py:132-135` — `zoi_moves()` has O(lookback × candidates) complexity

```python
for q0, r0 in recent:
    for cq, cr in self.candidates:
        if _hex_dist(q0, r0, cq, cr) <= margin:
            within.add((cq, cr))
```

This is O(L × C) where L=lookback and C=|candidates|. In late-game positions with hundreds of candidates, this becomes a bottleneck called on every MCTS simulation. The `_hex_dist` calculation involves 5 arithmetic operations per pair.

**Recommendation:** Use a spatial hash (bin candidates by (q÷margin, r÷margin)) for O(L) lookups, or precompute the ZOI set once per move rather than per simulation.

### 8. `inference.py:245` — Redundant board array duplication in `_process_batch`

```python
for board_arr_p, p in zip([board_arr] * n, planes):
    board_rows.append(board_arr_p)
```

Each request's board array is appended N times (once per move). For a request with 80 legal moves, this creates 80 references to the same numpy array in the list but then `np.stack()` materializes all copies. With `BOARD_SIZE=18` and `IN_CH=11`, that's 80×11×18×18×4 bytes = ~1.1MB per request, stacked redundantly.

**Recommendation:** The trunk already processes each board once (line 297-298). The redundancy is in `boards_t` not in the forward pass itself. Consider separating the board batch (one per request) from the move batch (many per request) to avoid the memory overhead.

### 9. No learning rate scheduling

The optimizer uses a fixed `LR=1e-3` throughout training. AlphaZero and KataGo both use cosine or step decay schedules. Without LR decay, the network may oscillate around minima in later generations.

**Recommendation:** Add `torch.optim.lr_scheduler.CosineAnnealingLR` or a step-based schedule.

---

### 10. Module-level constants read from CFG are frozen at import time

```python
# net.py:34-35
HIDDEN   = CFG["TRUNK_CHANNELS"]
N_BLOCKS = CFG["TRUNK_BLOCKS"]
```

And in `train.py`:
```python
BATCH_SIZE = CFG["BATCH_SIZE"]
LR = CFG["LR"]
```

These are set at import time and never re-read. If `config.py` is updated via the dashboard (`api_config_write`), the running process still uses the old values. The dashboard gives users the false impression that config changes take effect on "next training run" but since `train.py` reads CFG once at import, even restarting via the dashboard subprocess may import the cached module.

**Recommendation:** Read CFG values at the start of `train()` rather than at module level, or use a config reload mechanism.

### 11. No gradient accumulation for small effective batch sizes

With `BATCH_SIZE=64` and the per-sample policy loop (issue #6), the effective batch for gradient computation is just 1 for the policy head. This can cause noisy gradients especially early in training.

**Recommendation:** If the per-sample loop is kept, accumulate gradients across multiple mini-batches before stepping.

### 12. Replay buffer has no prioritization

The buffer is a simple FIFO deque with recency weighting. Prioritized experience replay (sampling proportional to TD error or loss) would improve sample efficiency, especially for rare but informative positions like near-win/near-loss states.

### 13. `test_game.py:446` — Typo catches wrong exception

```python
except (AssertionError, Exception) as e:
```

`AssertionError` is not a real exception class — it's a typo for `AssertionError`. However, since `Exception` is also caught (and `AssertionError` is a subclass of `Exception`), this doesn't cause a runtime failure. But the explicit mention of `AssertionError` is dead code.

**Fix:** Change to `except Exception as e:` or fix the typo to `AssertionError`.

### 14. No test framework — custom test runner

`test_game.py` uses a hand-rolled test runner instead of pytest (which is even listed in `requirements.txt`). This means no parallel execution, no fixtures, no parameterization, and no CI integration out of the box.

**Recommendation:** Convert to standard pytest tests (just remove the `if __name__` runner — the functions already follow pytest naming conventions).

---

### 15. Inconsistent device handling

`net.py` uses `DEVICE` (a module-level `torch.device`). `inference.py` has its own `_CUDA` flag. `train.py` has another `_CUDA` flag. The CUDA Graph path in `inference.py:118` hardcodes `device="cuda"` instead of using `DEVICE`.

**Recommendation:** Single source of truth for device selection, imported everywhere.

### 16. `dashboard.html` — No input sanitization on replay filenames

```javascript
btn.onclick = () => loadReplay(name);
```

While the server validates filenames (no `/`, `\`, `..`), the client passes the raw server response directly into a fetch URL. If the server response were ever tampered with, this could lead to unexpected requests.

### 17. SSE reconnection has no backoff limit

```javascript
es.onerror = () => { es.close(); setTimeout(connectSSE, 3000); };
```

Fixed 3-second retry with no exponential backoff or max-retry limit. If the server is down, this creates a tight reconnection loop indefinitely.

### 18. `replay.py` duplicates 1-2-2 turn logic

The turn-switching logic in `replay.py:58-62` manually reimplements the same logic as `HexGame.make()`. Any future rule changes would need to be updated in both places.

**Recommendation:** Import and use `HexGame` directly to replay moves.

### 19. Binary replay files committed to git

The `replays/` directory contains `.json` replay files and `.png` heatmap images that are generated artifacts. These will grow unboundedly with training. The `.gitignore` ignores some artifacts but not `replays/`.

**Recommendation:** Add `replays/*.json` and `replays/*.png` to `.gitignore`, or move to a separate data branch.

### 20. `requirements.txt` pins exact versions including CUDA-specific torch

```
torch==2.6.0+cu124
```

This will fail on machines without CUDA 12.4 or on macOS/CPU-only setups. Standard practice is to list `torch>=2.0` and let users install the appropriate CUDA variant separately.

---

## Positive Highlights

These are things done well that are worth calling out:

- **D6 symmetry augmentation** (`net.py:49-161`): Mathematically rigorous implementation of the full dihedral group of the hexagonal lattice. The 12-element group is correctly enumerated with both rotations and reflections, and consistently applied to boards, moves, and auxiliary labels.

- **HexConv2d kernel masking** (`net.py:241-266`): Clever structural prior that enforces the Z[omega] neighbourhood geometry at the convolution level rather than relying on the network to learn it.

- **Make/unmake pattern** (`game.py:42-99`): Zero-allocation tree traversal for MCTS. The undo stack correctly tracks all state (candidates, placements, winner, player) for perfect rollback.

- **Batched inference server** (`inference.py`): Well-designed producer-consumer pattern with thread-safe queues, persistent caching, and CUDA Graph support. Solves a real performance problem (sequential GPU calls) with a clean abstraction.

- **Cosine temperature annealing** (`train.py:300`): Smooth exploration→exploitation transition that avoids the sharp phase transitions of step-based schedules.

- **Comprehensive test suite** (`test_game.py`): 26 tests covering the game engine, board encoding, D6 symmetry properties (rotation composition, reflection involution), and agent behavior.

- **Dashboard** (`dashboard.html` + `server.py`): Full-featured single-file training dashboard with real-time SSE metrics, replay visualization, and config editing.

---

## Summary Table

| # | Severity | File | Issue |
|---|----------|------|-------|
| 1 | Critical | server.py | `exec()` on config = RCE risk |
| 2 | Critical | inference.py | CUDA Graph value index bug |
| 3 | Critical | train.py | Aux losses recompute trunk |
| 4 | Critical | mcts.py | Child player assignment in 1-2-2 |
| 5 | Critical | tune.py | Hardcoded Windows Python path |
| 6 | High | train.py | Per-sample policy loss loop |
| 7 | High | game.py | O(L×C) ZOI pruning |
| 8 | High | inference.py | Redundant board duplication |
| 9 | High | train.py | No LR scheduling |
| 10 | Medium | net.py/train.py | CFG frozen at import time |
| 11 | Medium | train.py | No gradient accumulation |
| 12 | Medium | train.py | No prioritized replay |
| 13 | Medium | test_game.py | AssertionError typo |
| 14 | Medium | test_game.py | No pytest integration |
| 15 | Low | multiple | Inconsistent device handling |
| 16 | Low | dashboard.html | Client filename sanitization |
| 17 | Low | dashboard.html | SSE reconnect no backoff |
| 18 | Low | replay.py | Duplicated turn logic |
| 19 | Low | .gitignore | Replay artifacts in git |
| 20 | Low | requirements.txt | CUDA-pinned torch version |

---

*This review was performed by reading every source file in the repository. Happy to discuss any of these points or help implement fixes.*
